crates/sui-rosetta/src/account.rs

```diff
@@ -0,0 +1,97 @@
+// Copyright (c) 2022, Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+use std::sync::Arc;
+
+use axum::{Extension, Json};
+
+use sui_core::authority::AuthorityState;
+use sui_types::base_types::SuiAddress;
+use sui_types::gas_coin::GasCoin;
+use sui_types::object::Owner;
+
+use crate::errors::Error;
+use crate::types::{
+    AccountBalanceRequest, AccountBalanceResponse, AccountCoinsRequest, AccountCoinsResponse,
+    Amount, Coin, CoinID, CoinIdentifier, SignedValue,
+};
+use crate::{ErrorType, OnlineServerContext, SuiEnv, SUI};
+
+pub async fn balance(
+    Json(payload): Json<AccountBalanceRequest>,
+    Extension(context): Extension<Arc<OnlineServerContext>>,
+    Extension(env): Extension<SuiEnv>,
+) -> Result<AccountBalanceResponse, Error> {
+    env.check_network_identifier(&payload.network_identifier)?;
+
+    let block_id = if let Some(index) = payload.block_identifier.index {
+        context.blocks().get_block_by_index(index).await.ok()
+    } else if let Some(hash) = payload.block_identifier.hash {
+        context.blocks().get_block_by_hash(hash).await.ok()
+    } else {
+        None
+    }
+    .map(|b| b.block.block_identifier);
+
+    if let Some(block_identifier) = block_id {
+        context
+            .blocks()
+            .get_balance_at_block(payload.account_identifier.address, block_identifier.index)
+            .await
+            .map(|balance| AccountBalanceResponse {
+                block_identifier,
+                balances: vec![Amount::new(balance.into())],
+            })
+    } else {
+        let gas_coins = get_coins(&context.state, payload.account_identifier.address).await?;
+        let amount: u128 = gas_coins.iter().map(|coin| coin.amount.value.abs()).sum();
+        Ok(AccountBalanceResponse {
+            block_identifier: context.blocks().current_block_identifier().await?,
+            balances: vec![Amount::new(amount.into())],
+        })
+    }
+}
+
+pub async fn coins(
+    Json(payload): Json<AccountCoinsRequest>,
+    Extension(context): Extension<Arc<OnlineServerContext>>,
+    Extension(env): Extension<SuiEnv>,
+) -> Result<AccountCoinsResponse, Error> {
+    env.check_network_identifier(&payload.network_identifier)?;
+    let coins = get_coins(&context.state, payload.account_identifier.address).await?;
+    Ok(AccountCoinsResponse {
+        block_identifier: context.blocks().current_block_identifier().await?,
+        coins,
+    })
+}
+
+async fn get_coins(state: &AuthorityState, address: SuiAddress) -> Result<Vec<Coin>, Error> {
```

Comment:
1, Currently the rosetta online server is a fullnode server, currently it's best to have local access to AuthorityState because we are fabricating blocks and computing historical balances locally. Once we have fullnode checkpoint query and get balance support, we can consider a "remote" mode where the rosetta server talks to the fullnode via RPC.

2, From our understanding the rosetta server is being use by CB for account balance reconciliation, so performance is not very important for now, also there are plan to add get coin and get balance in the RPC API, we can replace this once it's implemented.